api/v1/autostoragescale.go

```diff
@@ -0,0 +1,75 @@
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// AutoStorageScalingSpec defines the desired state of AutoStorageScaling
+type AutoStorageScalingSpec struct {
+	// StorageClusterName is the name of the storage cluster for which the storage scaling is to be done.
+	// +kubebuilder:validation:Required
+	StorageClusterName string `json:"storageClusterName,omitempty"`
+	// DeviceClassName is the name of the device class for which the storage scaling is to be done.
+	// +kubebuilder:validation:Required
+	DeviceClassName string `json:"deviceClassName,omitempty"`
+	// CapacityLimit is the capacity limit for the storage scaling for the specific deviceClass and storagecluster.
+	// +kubebuilder:validation:Required
+	CapacityLimit resource.Quantity `json:"capacityLimit,omitempty"`
+}
+
+// AutoStorageScalingStatus defines the observed state of AutoStorageScaling
+type AutoStorageScalingStatus struct {
+	// Phase describes the Phase of AutoStorageScaling
+	Phase string `json:"phase,omitempty"`
+	// LastRunTimeStamp is the time stamp of the last run of the storage scaling
+	LastRunTimeStamp metav1.Time `json:"lastRunTimeStamp,omitempty"`
+	// The known size is the original size of the OSDs before the expansion in progress is completed.
+	// After the expansion is completed, this would be updated to the expected OSD size.
+	// Used for vertical scaling of OSDs.
+	KnownOsdSize resource.Quantity `json:"knownOsdSize,omitempty"`
+	// The ExpectedOsdSize is the size that the auto-expansion has decided to set.
+	// This will be set on the storageCLuster CR as the desired size of the OSDs.
+	// Used for vertical scaling of OSDs.
+	ExpectedOsdSize resource.Quantity `json:"expectedOsdSize,omitempty"`
+	// The known OSD count is the original count of the OSDs before the expansion in progress is completed.
+	// After the expansion is completed, this would be updated to the expected OSD count.
+	// Used for horizontal scaling of OSDs.
+	KnownOsdCount int32 `json:"knownOsdCount,omitempty"`
+	// The Expected OSD count is the count that the auto-expansion has decided to set.
+	// This will be set on the storageCLuster CR as the desired count of the OSDs.
+	// Used for horizontal scaling of OSDs.
+	ExpectedOsdCount int32 `json:"expectedOsdCount,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=.metadata.creationTimestamp, description="Age"
+// +kubebuilder:printcolumn:name="Phase",type=string,JSONPath=.status.phase,description="Current Phase"
+// +kubebuilder:printcolumn:name="LastRunTimeStamp",type=date,JSONPath=.status.lastRunTimeStamp,description="Last Run Time Stamp"
+// +operator-sdk:csv:customresourcedefinitions:displayName="Auto Storage Scaling"
+
+// AutoStorageScaling represents the automatic storage scaling for odf cluster.
```

Comment:
how about calling it as `StorageAutoScaling` or `StorageAutoScaler` ?  because keeping Storage in the front aligns with the idea that we are trying to scale storage.

Reply:
+1

Reply:
Updated